### ISSUE TITLE
[blocked on #20] Try sparser and not ad-hoc data structures for the scaling factors

### ISSUE THREAD

**Issue by Mikolaj** [Mikolaj/horde-ad#18]
Currently we just use vectors for the scaling factor in `eval1`, which is too dense and so not optimal, `MatrixOuter` for `eval2`, which is ad hoc, and tensors for `evalX` and `evalS`, which are monstrous. Perhaps for the scaling factor in `eval_n+1` just use `Delta_n` or a close analogue?

[Edit: a fix of most of #13 now unblocked this. Then it turned out #20 blocks it still.] This ticket is currently blocked on #13 that would let us use other compiler versions than GHC 9.2 and #14 that would fix degraded and unpredictable performance in 9.2 series (and HEAD), perhaps as soon as in GHC 9.2.3. It's too easy to break performance in this code-base and too hard to detect it and bisect the culprit if measurements are unreliable for huge portions of the repo history. That's particularly important when implementing optimizations.

Related scribbles copied from the google doc:

```Haskell
eval1 :: Delta1 r -> Delta1 r -> ST s ()

eval1 :: VectorRecipe r -> Delta1 r -> ST s ()
data VectorRecipe r where
   vrScale :: Float -> r -> r
   ...
```

**Comment by Mikolaj** [comment]
This may be no longer relevant, but I can't tell.

See also this bit noted by @tomjaguarpaw: "structurally linear functions are a representation for matrices. Indeed, they are a maximally expressive representation for matrices, that can encode any structured (or unstructured) sparsity pattern whatever" in the YOLO paper https://arxiv.org/pdf/2204.10923.pdf. This would point to using Delta as a sparse representation for the "scaling factor" (now called by us "cotangent accumulator"). Another angle: in the Ast-gradient-producing mode we use tensor Ast terms as cotangent accumulators, so this is already sparse in a similar sense and, moreoever, Ast is not linear and so it's able to express the non-linear coefficients that need to be present, even though they are fixed for any single gradient descent step.

**Comment by Mikolaj** [comment]
I do think the current use of AST as the cotangent contribution is as sparse as it gets, so this ticket can be closed.
